### ISSUE TITLE
shapviz::shapviz: Error in s[, nms, drop = FALSE] : incorrect number of dimensions

### ISSUE THREAD

**Issue by sebsilas** [ModelOriented/shapviz#141]
When I run `shapviz::shapviz`, I get the error:

> Error in s[, nms, drop = FALSE] : incorrect number of dimensions

I believe I am using the correct predictors for the model, which are being "retained" for me by `recipes::bake`.

I'm also removing NAs, just incase.

Any idea what is going on here?

```{r}

syllable_mod <- bundle::unbundle(lyricassessr::syllable_classifier_bundle)

# The tidymodels prediction works:

  preds <- predict(syllable_mod, new_data = audio_features, type = "prob") %>%
    dplyr::rename_with(~stringr::str_remove_all(.x, ".pred_")) %>%
    tidyr::pivot_longer(dplyr::everything(),
                        names_to = "Syllable", values_to = "Probability") %>%
    dplyr::arrange(dplyr::desc(Probability))

  xgb_model <- parsnip::extract_fit_engine(syllable_mod)

# Make sure there are no NAs, just incase (1 is a meaningless number for now..)
  audio_features <- audio_features %>%
    dplyr::mutate(across(everything(), ~ ifelse(is.na(.x), 1, .x)))

  audio_features_prepped <- recipes::bake(
    lyricassessr::prepped_recipe,
    recipes::has_role("predictor"),
    new_data = audio_features) %>%
    dplyr::mutate(across(everything(), ~ ifelse(is.na(.x), 1, .x)))

# Make sure audio_features agree with the retained (prepped) features
  audio_features <- audio_features %>%
    dplyr::select(dplyr::all_of(names(audio_features_prepped)))


shap_values <- shapviz::shapviz(xgb_model, X_pred = as.matrix(audio_features_prepped), X = as.matrix(audio_features))

```

**Comment by mayer79** [comment]
Don't pass `X`, maybe?

**Comment by sebsilas** [comment]
> Which XGB version are you using?

1.7.7.1

**Comment by mayer79** [comment]
I can't run your code, so its all about speculation.

You can check what  

```r
s <- predict(xgb_model, X_pred = as.matrix(audio_features_prepped, predcontrib = TRUE)
```
The result `s` should be a list with matrices of dimension (nrow(X_pred), num_classes).

**Comment by mayer79** [comment]
This should work:

```r
library(shapviz)
library(xgboost)

params <- list(objective = "multi:softprob", num_class = 3)
X_pred <- data.matrix(iris[, -5])
dtrain <- xgboost::xgb.DMatrix(X_pred, label = as.integer(iris[, 5]) - 1)
fit <- xgb.train(params = params, data = dtrain, nrounds = 10)

x <- shapviz(fit, X_pred = X_pred)
x
sv_importance(x)
sv_dependence(x$Class_3, v = colnames(iris[, -5]))
```
<img width="361" alt="image" src="https://github.com/user-attachments/assets/a4eb19bf-cfe8-4755-8cc8-c224608bfd93">

<img width="675" alt="image" src="https://github.com/user-attachments/assets/c56a662c-354c-4432-bbeb-636922e0afa1">

**Comment by sebsilas** [comment]
> I can't run your code, so its all about speculation.
> 
> You can check what
> 
> ```r
> s <- predict(xgb_model, X_pred = as.matrix(audio_features_prepped), predcontrib = TRUE)
> ```
> 
> gives. The result `s` should be a list with num classes matrices of dimension (nrow(X_pred), ncol(X_pred) + 1).


Yes, (if I change `X_pred` to `newdata`), I get a list with length == number of classes and matrices of the dimensions you specified (+1 being the `BIAS` column).

**Comment by sebsilas** [comment]
> This should work:
> 
> ```r
> library(shapviz)
> library(xgboost)
> 
> params <- list(objective = "multi:softprob", num_class = 3)
> X_pred <- data.matrix(iris[, -5])
> dtrain <- xgboost::xgb.DMatrix(X_pred, label = as.integer(iris[, 5]) - 1)
> fit <- xgb.train(params = params, data = dtrain, nrounds = 10)
> 
> x <- shapviz(fit, X_pred = X_pred)
> x
> sv_importance(x)
> sv_dependence(x$Class_3, v = colnames(iris[, -5]))
> ```
> 
> <img alt="image" width="361" src="https://private-user-images.githubusercontent.com/16206095/357436867-a4eb19bf-cfe8-4755-8cc8-c224608bfd93.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM1NTg4ODgsIm5iZiI6MTcyMzU1ODU4OCwicGF0aCI6Ii8xNjIwNjA5NS8zNTc0MzY4NjctYTRlYjE5YmYtY2ZlOC00NzU1LThjYzgtYzIyNDYwOGJmZDkzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODEzVDE0MTYyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZjMDYwOTAzMDlkZGIxY2U4MzY1MGJhYThjMjJmMDQxY2IxMTM3OGU0MWJlZmRhNTNkMjA2ZWYwZTQ1MTg2Y2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.vsRGRpGqD829_pOo2kcbFciIRK_KZ-P379PwigdojaI"> <img alt="image" width="675" src="https://private-user-images.githubusercontent.com/16206095/357436731-c56a662c-354c-4432-bbeb-636922e0afa1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM1NTg4ODgsIm5iZiI6MTcyMzU1ODU4OCwicGF0aCI6Ii8xNjIwNjA5NS8zNTc0MzY3MzEtYzU2YTY2MmMtMzU0Yy00NDMyLWJiZWItNjM2OTIyZTBhZmExLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODEzVDE0MTYyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgxMjBjMmFkNGI4Y2IxZDAyNjk3NTllYWNiZTljOWY4ODE4MzM1ZDBjM2QxYmU3ZmQxODMxMTczM2JlOTVhNjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.4vUnsaJxlB3urqkSmebbillbDTFNLWLTs6_Rhs_ViOA">

This does work for me. But I am fitting the model via `tidymodels` then extracting it. 

I tried replicating [this](https://www.r-bloggers.com/2023/01/shap-xgboost-tidymodels-love/) which fits the xgboost model via `tidymodels`, as much as I could too, but I still have an issue.

**Comment by mayer79** [comment]
Not sure...

```r
library(shapviz)

audio_features <- bundle::unbundle(lyricassessr::vowel_metadata_with_audio_features_range_restricted_without_AEIOU)
# run your code
shap_values <- shapviz(xgb_model, X_pred = as.matrix(audio_features_prepped))
sv_importance(shap_values)
```

![image](https://github.com/user-attachments/assets/0c3cdd8c-178e-481f-9faa-f9f2589bd171)

**Comment by sebsilas** [comment]
Your code doesn't appear to be right

> Not sure...
> 
> ```r
> library(shapviz)
> packageVersion("shapviz") # ‘0.9.3’
> 
> audio_features <- bundle::unbundle(lyricassessr::vowel_metadata_with_audio_features_range_restricted_without_AEIOU)
> # run your code
> shap_values <- shapviz(xgb_model, X_pred = as.matrix(audio_features_prepped))
> sv_importance(shap_values)
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/16206095/357478155-0c3cdd8c-178e-481f-9faa-f9f2589bd171.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM1NjQ0MDcsIm5iZiI6MTcyMzU2NDEwNywicGF0aCI6Ii8xNjIwNjA5NS8zNTc0NzgxNTUtMGMzY2RkOGMtMTc4ZS00ODFmLTlmYWEtZjlmMjU4OWJkMTcxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODEzVDE1NDgyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEwYjVlNGQyMWRmOWNjMGZmMjU2MTJkNWZkY2NhM2E0MzVlZDRjNTdlZjM5YTZiYzQ2NWE2MjE0OWQwZWUxMzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Yck6d-rwV8AgmEHBDsuMzYzMRt5tYjMngK2dZRQNbog)

Hm, well that can't be right. `bundle::unbundle` is meant to unbundle a serialized model object (see [here](https://rstudio.github.io/bundle/)).

How are you defining `xgb_model` and `audio_features_prepped` in your example?

**Comment by mayer79** [comment]
Both according to your code. But since your code lacks the definition of `audio_features`, I took the first object that would not produce an error.^^

**Comment by sebsilas** [comment]
Right okay, I can reproduce this now.

I think I have maybe misunderstood what you can do with Shapley values. For some reason, I thought for a *new* prediction, you could work out which variables are contributing most to a given prediction. But actually, you can only do this with the full training dataset, right? So I get the same error when I update your example to e.g., `slice_sample` one row off the training data. It only works with the full dataset.. my bad!

**Comment by mayer79** [comment]
If you give me "audio_features", I can probably make things run correctly. Currently, I am not 100% certain if XGBoost gets the right input. It was trained on 14 features, but we are passing 31.

**Comment by mayer79** [comment]
You can use any plot, e.g.:

```r
sv <- shapviz(xgb_model, X_pred = audio_features_prepped[1:2, , drop = FALSE])
sv_waterfall(sv$Class_1, row_id = 1)
```
![image](https://github.com/user-attachments/assets/a547f1d5-c8f8-40ed-adfb-1cb7af401f34)

Actually, when I only pass a single row to X_pred, I get your error message.

**Comment by sebsilas** [comment]
Ah, so this is the issue then: I am only ever using 1 row!

**Comment by mayer79** [comment]
- I will fix this on shapviz's side
- Actually, XGBoost is also working on a big update for R, which solves this issue. But it might take some time until the release is here.
- As a quick fix, you can simply replicate your row a second time.

**Comment by sebsilas** [comment]
Ok, many thanks. Do let me know when you've pushed a fix and I'll be happy to test it.

I assume replicating the row will not effect the estimates in any biased way?

**Comment by sebsilas** [comment]
By the way, as a separate point, [this link](https://www.r-bloggers.com/2023/01/shap-xgboost-tidymodels-love/) says that if you pass the original data structure to the `X` argument, you will get labelled classes, but I can't replicate that here (I get Class_1, Class_2 etc. as in your screenshot):

```
shap_values <- shapviz(xgb_model, X_pred = as.matrix(audio_features_prepped), X = as.matrix(audio_features))

sv_importance(shap_values)
```

Should I be able to get the proper class labels (how if so?)

**Comment by mayer79** [comment]
XGB loses the class labels. But you can set them as `names(shapviz_object) <- ....` (like a list)

**Comment by sebsilas** [comment]
Hm ok, I hope this is a safe way of doing it (i.e., guaranteeing the correct assignment)!

```
outcome_classes <- audio_features$syllable %>% as.character() %>%  unique()
names(shap_values) <- outcome_classes
```

**Comment by mayer79** [comment]
It must be in the same order as tidymodels does the class assignment when creating the response for the XGBoost backend. I let you figure this out. It is not under my control.

**Comment by mayer79** [comment]
I think it should be `levels(train_y)` (if factor), or `levels(as.factor(train_y))`.
